### ISSUE TITLE
Angular Issue

### ISSUE THREAD

**Issue by ksr583** [kentcdodds/ama#49]
I am having a very odd issue on an angular app I'm writing for a client, and I was hoping you would have some insight that could help me track down and fix it.

The app is a typical CRUD app that helps manage entities that, for the sake of this discussion we'll call Clients, Locations, and Employees. The data is pulled off of an API and displayed/manipulated from there. The issue I'm having is this: When you navigate to the "listings" page for any of the entities, the data is pulled from the API via resolve (using ui-router) and bound to the view in the controller. In the template, the list is rendered using ng-repeat. When you navigate to the list page via the menu, or any other UI element that points to it, it renders just fine. However, if you are on the list, click an entity (which takes you to the details page for that entity), and then click the back button in the browser, the list page takes ~25s to render. It is repeatable every time.

Here are some of the steps I've taken, none of which made any difference:
- I've tried limiting the list data to fewer entries (as low as 10)
- I've ensured that the response time from the API is consistent
- I have tried adding in one-time bindings (e.g. `{{ ::client.name }}`), and using `track by`.
- I installed your ng-stats snippet and run it. The watchers are consistent whether you navigate to the page via UI or the back button, but the digest times are vastly different: ~565 digest via UI, and ~22151 via the browser back button.
- I've tried temporarily reducing the number of data points that are rendered for each list item in order to reduce the number of watchers overall.

I recorded the transition via the back button using Dev Tools, and [here is a screenshot](https://www.dropbox.com/s/yglzmj4vf7kuwlt/Screenshot%202015-11-29%2011.44.45.png?dl=0) of the aggregated details of the process that took all of the time.

Any ideas you have would be most appreciated!

**Comment by kentcdodds** [comment]
Here are a few things you might try:

- use [ng-stats](https://github.com/kentcdodds/ng-stats) to see digest times
- Try it in an incognito tab (so it doesn't have your extensions). It looks like you have Batarang installed and enabled. Could be that...
- Open up the devtools and click the "Pause" button to see where execution is.

Pause button:

![screen shot 2015-11-30 at 9 45 51 am](https://cloud.githubusercontent.com/assets/1500684/11477870/2d961e58-9747-11e5-8af3-0ca8ffcff755.png)

Let me know what you come up with :-)

**Comment by ksr583** [comment]
I had already been using ng-stats to try and track down the problem with no luck (although it did help me improve other things about the app by reducing watchers via one-way binding. In the end I resorted to process of elimination: removing things from my controller and routing until the problem went away. 

In the end it turned out the culprit was `angular.copy()`. I found [this thread](https://github.com/angular/angular.js/issues/11099) where one of the Angular collaborators even said that the team regrets exposing angular.copy to the public, and that led me to removing it from my controller to test – I had about 6-7 of them in the offending controller. Voila, page load time went back to normal (~1s). 

I can't remember exactly why I was using angular.copy in that situation, as everything still worked the same after I removed it, but I think it had something to do with sorting/manipulating lists off of an API by making a copy of them first. Anyway, issue resolved!
